### ISSUE TITLE
TargetParameterCountException when using event Raise with Moq

### ISSUE THREAD

**Issue by PavelReichl** [devlooped/moq#1285]
I have a class with following event:

```csharp
_event Func<ApplicationMessageProcessedEventArgs, Task> ApplicationMessageProcessedAsync;
```

I would like to test the handler for an event and have following code_

```csharp
mqttClient.Mock.Raise(client => client.ApplicationMessageReceivedAsync += null, eventArg);
```

I am getting exception:

```
System.Reflection.TargetParameterCountException: Parameter count mismatch.
    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Moq.Extensions.InvokePreserveStack(Delegate del, IReadOnlyList`1 args) in C:\projects\moq4\src\Moq\Extensions.cs:line 158
   at Moq.Mock.RaiseEvent(Mock mock, LambdaExpression expression, Stack`1 parts, Object[] arguments) in C:\projects\moq4\src\Moq\Mock.cs:line 745
   at Moq.Mock.RaiseEvent[T](Mock mock, Action`1 action, Object[] arguments) in C:\projects\moq4\src\Moq\Mock.cs:line 695
   at Moq.Mock`1.Raise(Action`1 eventExpression, EventArgs args) in C:\projects\moq4\src\Moq\Mock`1.cs:line 1366
   at MqttChannel.MqttChannelTest.MqttChannel_ProcessNewMessage() in C:\Development\Shared\OT.Test\MqttChannel\MqttChannelTest.cs:line 298
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadOperations.ExecuteWithAbortSafety(Action action)
```

Any idea how to solve this?

**Comment by stakx** [comment]
Not without some repro code. The following works perfectly fine:

```csharp
using Moq;

var mock = new Mock<IFoo>();
var eventArgs = new ApplicationMessageProcessedEventArgs();
mock.Raise(m => m.ApplicationMessageProcessedAsync += null, eventArgs);

public class ApplicationMessageProcessedEventArgs : EventArgs { }

public interface IFoo
{
	event Func<ApplicationMessageProcessedEventArgs, Task> ApplicationMessageProcessedAsync;
}
```

So obviously your code differs in some way. Without knowing how, there's no way to help you.

(Please note that this issue tracker is for the development of Moq, and not so much for plain usage questions. So unless you think you've found a bug, you'd be better off asking this in places such as Stack Overflow. You'll likely have a much larger audience there than here.)

**Comment by PavelReichl** [comment]
[MoqSampeTest.zip](https://github.com/moq/moq4/files/9470084/MoqSampeTest.zip)

Prepared repro code.

**Comment by stakx** [comment]
I'll reproduce a slightly adjusted excerpt from your repro code so it's more readily understandable without having to look at the whole thing:

```csharp
public interface IMqttClient
{
    event Action<EventArgs> ApplicationMessageAsync;
    event EventHandler ApplicationMessage;
}

// Mock<IMqttClient> mqttClientMock = ...;
mqttClientMock.Raise(client => client.ApplicationMessage += null, mqttClient, EventArgs.Empty);
mqttClientMock.Raise(client => client.ApplicationMessageAsync += null, EventArgs.Empty);
```

You're making a few mistakes here.

```csharp
mqttClient.Mock.Raise(client => client.ApplicationMessageAsync += null, EventArgs.Empty);
```

You're passing a single `EventArgs` argument here after the `Raise` expression, so you're using the method overload		`mock.Raise(Action<T> eventExpression, EventArgs args)`. This overload is intended to be used with classic .NET `EventHandler` events, and it automaticaly adds `mock.Object` as the `sender` argument. Thus your event handlers are called with _two_ arguments, `sender := mqttClientMock.Object` (automatically added), and `e := EventArgs.Empty` (specified by you). And since `ApplicationMessageAsync` only accepts one argument, you get the exception.

```csharp
mqttClientMock.Raise(client => client.ApplicationMessage += null, mqttClient, EventArgs.Empty);
```

Here, you're trying to get the overload specifically for classic `EventHandler` events, but because you're explicitly passing the `sender` argument, you don't end up in the right overload for it.

Your code would have been more correct as follows:

```diff
-mqttClientMock.Raise(client => client.ApplicationMessage += null, mqttClient, EventArgs.Empty);
+mqttClientMock.Raise(client => client.ApplicationMessage += null, EventArgs.Empty);
```

... because `sender` is provided automatically.

```diff
-mqttClientMock.Raise(client => client.ApplicationMessageAsync += null, EventArgs.Empty);
+mqttClientMock.Raise(client => client.ApplicationMessageAsync += null, (object)EventArgs.Empty);
```

... because you don't want to end up in the `Raise` method overload geared towards `EventHandler` events.

Now, as to the question of how to properly call async events, this is something that Moq doesn't currently have full support for. I mentioned in https://github.com/moq/moq4/issues/977#issuecomment-1193292475:

> We could in theory add a `mock.RaiseAsync` method specifically for events whose handlers return a `Task`. `Raise` would then invoke each handler, gather its `Task`, and either `await` them in series or in parallel (using `Task.WhenAll`).

Such a `RaiseAsync` method has not been implemented so far, but we can look at that in a separate issue.
